### PR TITLE
Handle unlistening from events synchronously from their handlers

### PR DESCRIPTION
**Pull request by chrisinajar**
This patch allows developers to call the unlisten method synchronously from an event handler and have things work the way one would expect.

Before this, if you unlisten from within that same handler then the final handler in the listener list will not fire.

Check the added test cases for the specific corner cases handled by this patch.

### CODE REVIEW

**Comment by chrisinajar** [comment]
Closes: https://github.com/Raynos/geval/issues/4

File: event.js [hunk 1]
```diff
@@ -2,12 +2,15 @@ module.exports = Event
 
 function Event() {
     var listeners = []
+    var listenersToBroadcast = []
```
Comment (Raynos) [inline event.js]: I don't like having a weird second variable in scope.

File: event.js [hunk 2]
```diff
@@ -21,6 +24,11 @@ function Event() {
             if (index !== -1) {
                 listeners.splice(index, 1)
             }
+            // if we're mid-broadcast then remove handlers that have not yet fired
+            index = listenersToBroadcast.indexOf(listener)
```
Comment (Raynos) [inline event.js]: two removals is going to get weird.

**Review by Raynos — COMMENTED** [review]
Consider an approach like:

```
function Event() {
    var listeners = []

    return { broadcast: broadcast, listen: event }

    function broadcast(value) {
        var listenersCopy = listeners.slice();
        for (var i = 0; i < listeners.length; i++) {
            if (!listenersCopy[i].deleted) {
                listenersCopy[i].fn(value)
            }
        }
    }

    function event(listener) {
        listeners.push(new ListItem(listener))

        return removeListener

        function removeListener() {
            for (var i = 0; i < listeners.length; i++) {
                if (listeners[i].fn === listener) {
                    listeners[i].deleted = true
                    listeners.splice(i, 1)
                    break
                }
            }
        }
    }
}

function ListItem(fn) {
    this.fn = fn;
    this.deleted = false;
}
```

Rather then having two lists, have a list of objects that keep track of whether they are "alive".

We still need to make a copy of the list when looping to avoid "skipping" items in the iterator due to splice() call.

**Comment by chrisinajar** [comment]
I fixed the 1 typo in your suggested fix and changed this PR to match that. It passes all the test cases I added!

**Comment by Raynos** [comment]
This is great, appreciate all the tests.

**Comment by Raynos** [comment]
Published 2.2.0
